rubi/rubi/data/sources/market.py

```diff
@@ -1,485 +1,164 @@
+        :type buy_gem: str
+        :param open: whether or not the offer is still active
+        :type open: bool
+        :param start_time: the timestamp of the earliest offer to return
+        :type start_time: int
+        :param end_time: the timestamp of the latest offer to return
+        :type end_time: int
+        :param first: the number of offers to return
+        :type first: int
+        :param order_by: the field to order the offers by (default: timestamp, options: timestamp, price) TODO: expand this list
+        :type order_by: str
+        :param order_direction: the direction to order the offers by (default: desc, options: asc, desc)
+        :return: a dataframe of offers placed on the market contract
+        :rtype: pd.DataFrame 
         """
 
         # set the offer entity 
-        Offer = self.rubicon_market_light.Offer
-        Take = self.rubicon_market_light.Take
-
-        # create the synthetic fields for the offer entity
-        Offer.pay_amt_formatted = Offer.pay_amt / 10 ** Offer.pay_gem.decimals
-        Offer.buy_amt_formatted = Offer.buy_amt / 10 ** Offer.buy_gem.decimals
-        Offer.paid_amt_formatted = Offer.paid_amt / 10 ** Offer.pay_gem.decimals
-        Offer.bought_amt_formatted = Offer.bought_amt / 10 ** Offer.buy_gem.decimals
-
-        Take.pay_amt_formatted = Take.pay_amt / 10 ** Take.pay_gem.decimals
-        Take.buy_amt_formatted = Take.buy_amt / 10 ** Take.buy_gem.decimals
-
-        # for each of the filters, add the filter to the where clause
-        # TODO: there is most likely a more elegant way to do the lower case conversion of the variables, probably using a decorator?
-        where = []
-        if maker:
-            maker = maker.lower()
-            where.append(Offer.maker == maker)
-        # TODO: this should be modified to return all offers for a given pair, not just the given direction
-        if taker: 
-            taker = taker.lower()
-            where.append(Offer.takes.taker.id == taker)
-        if pair:
-            pair[0] = pair[0].lower()
-            pair[1] = pair[1].lower()
-            where.append(Offer.pay_gem == pair[0])
-            where.append(Offer.buy_gem == pair[1])
-        if filled:
-            where.append(Offer.filled == filled)
-        if cancelled:
-            where.append(Offer.cancelled == cancelled)
-        if live:
-            where.append(Offer.live == live)
-        if pay_gem:
-            pay_gem = pay_gem.lower()
-            where.append(Offer.pay_gem == pay_gem)
-        if buy_gem:
-            buy_gem = buy_gem.lower()
-            where.append(Offer.buy_gem == buy_gem)
-        if start_time:
-            where.append(Offer.timestamp >= start_time)
-        if end_time:
-            where.append(Offer.timestamp <= end_time)
+        Offer = self.data.Offer
+
+        # set the order_by parameter
+        if order_by.lower() not in ('timestamp', 'price'):
```

Comment:
I think we should have a discussion around validations:

my two cents. It is always best to raise all validation errors at once so then when a user makes a broken call they see everything that is wrong with it.

The worst is when you make a call. It returns with an error. You fix that error and then retry the call and it fails again for a different reason.

A way to do this is run all the validations and add any error messages to a list. Then at the end of all the validations if the list is non-empty to concatenate all the strings and return it as one error.

Reply:
excellent idea and approach, will adopt it accordingly!